crates/runtime/src/instance/allocator/pooling.rs

```diff
@@ -0,0 +1,1714 @@
+        mut req: InstanceAllocationRequest,
+    ) -> Result<InstanceHandle, InstantiationError> {
+        let index = {
+            let mut free_list = self.free_list.lock().unwrap();
+            if free_list.is_empty() {
+                return Err(InstantiationError::Limit(self.max_instances as u32));
+            }
+            let free_index = strategy.next(free_list.len());
+            free_list.swap_remove(free_index)
+        };
+
+        let host_state = std::mem::replace(&mut req.host_state, Box::new(()));
+
+        unsafe {
+            debug_assert!(index < self.max_instances);
+            let instance =
+                &mut *(self.mapping.as_mut_ptr().add(index * self.instance_size) as *mut Instance);
+
+            instance.module = req.module.clone();
+            instance.offsets = VMOffsets::new(
+                std::mem::size_of::<*const u8>() as u8,
+                instance.module.as_ref(),
+            );
+            instance.host_state = host_state;
+
+            Self::set_instance_memories(
+                instance,
+                self.memories.get(index),
+                self.memories.max_wasm_pages,
+            )?;
+            Self::set_instance_tables(instance, self.tables.get(index), self.tables.max_elements)?;
+
+            initialize_vmcontext(instance, req);
+
+            Ok(InstanceHandle::new(instance as _))
+        }
+    }
+
+    fn deallocate(&self, handle: &InstanceHandle) {
+        let addr = handle.instance as usize;
+        let base = self.mapping.as_ptr() as usize;
+
+        debug_assert!(addr >= base && addr < base + self.mapping.len());
+        debug_assert!((addr - base) % self.instance_size == 0);
+
+        let index = (addr - base) / self.instance_size;
+        debug_assert!(index < self.max_instances);
+
+        unsafe {
+            // Decommit any linear memories that were used
+            for (mem, base) in (*handle.instance)
+                .memories
+                .values()
+                .zip(self.memories.get(index))
+            {
+                let size = (mem.size() * WASM_PAGE_SIZE) as usize;
+                if size > 0 {
+                    decommit(base, size);
+                }
+            }
+
+            // Decommit any tables that were used
+            let table_element_size = max_table_element_size();
+            for (table, base) in (*handle.instance)
+                .tables
+                .values()
+                .zip(self.tables.get(index))
+            {
+                let size = round_up_to_pow2(
+                    table.size() as usize * table_element_size,
+                    self.tables.page_size,
+                );
+                if size > 0 {
+                    decommit(base, size);
+                }
+            }
+
+            // Drop the host state
+            (*handle.instance).host_state = Box::new(());
+        }
+
+        {
+            self.free_list.lock().unwrap().push(index);
+        }
```

Comment:
I'll remove.  I think originally this had a named scope guard rather than having the single statement, so it was explicitly done to denote a lock in play.